Removed flow and replaced with source/sink and relative alignment (#139)
chipsalliance/firrtl-spec: spec.md

{word: UInt<32>, valid: UInt<1>, flip ready: UInt<1>}
```

#### Alignment: Flipped vs Aligned

A subcomponent's alignment is a relative property: a subcomponent is aligned/flipped relative to another subcomponent of the same component or type.
Hence, a subcomponent is flipped/aligned *with respect to (w.r.t)* another subcomponent of that type (parent, sibling, child etc.).

The following example of a non-nested bundle `Parent` is used to list all of the alignment relationships between members of `p`.

```
wire p: { alignedChild: UInt<32>, flip flippedChild: UInt<32> }
```

First, every subcomponent/component is always aligned with themselves:

- `p` is aligned w.r.t `p`
- `p.alignedChild` is aligned w.r.t `p.alignedChild`
- `p.flippedChild` is aligned w.r.t `p.flippedChild`

Next, all parent/child relationships:
Because the `flippedChild` field is `Flipped`, it changes its aligment relative to its parent.

- `p` is aligned w.r.t `p.alignedChild`

> - `p` is flipped w.r.t `p.flippedChild`

Reviewer: this is sort of weirdly written. I know it is a symmetrical relationship, but wouldn't it be more intuitive to write it as

`p.aligned child is aligned w.r.t. p`

### Define

Define statements are used to route references through the design, and may be used wherever is most convenient in terms of available identifiers -- their location is not significant other than scoping, and do not have last-connect semantics.
Every sink probe expression must be the target of exactly one of these statements.

> The define statement takes a sink static reference target and sets it to the specified reference, which must either be a compatible probe expression or static reference source.

Reviewer: ```suggestion
The `define`{.firrtl} statement takes a sink static reference target and sets it to the specified reference, which must either be a compatible probe expression or static reference source.
```

The tail operation truncates the n most significant bits from e.
n must be non-negative and less than or equal to the bit width of e.

# Source and Sink Expressions

An expression can either be used as either a source expression or a sink expression; if an expression is on the left-hand-side of a connect, it is a sink expression.
All other expressions are source expressions.

> All hardware components can be referenced with a source expression.

Reviewer 2: The work that @mmaloney-sf has been doing on the grammar and general allows this to be simplified.

```suggestion
All expressions may be used as source expressions.
```

As written, it's not specified if primops, muxes, and other things are legal as sources. And they are!

Author: Thanks!

With the following exceptions, all other hardware components/subcomponents can be referenced with a sink expression:
  - input ports (or subcomponents which are input ports)
  - data field of memory read ports

>   - rdata field of memory read-write ports

Reviewer 2: Additionally, no primitive operations may be used as sink expressions. Given @mmaloney-sf's new grammar, it should be safe to say that only `references` are legal sink expressions. Then introduce the restrictions on what references are not legal as sinks.

Author: Excellent, I'll revisit this when I bump.

The tail operation truncates the n most significant bits from e.
n must be non-negative and less than or equal to the bit width of e.

# Source and Sink Expressions

An expression can either be used as either a source expression or a sink expression; if an expression is on the left-hand-side of a connect, it is a sink expression.
All other expressions are source expressions.

All expressions may be used as source expressions.

With the following exceptions, all other hardware components/subcomponents can be referenced with a sink expression:

- input ports (or subcomponents which are input ports)
- data field of memory read ports

> - rdata field of memory read-write ports

Reviewer: if we use the nomenclature "ports" (as seen from within a module) and "pins" (as seen from outside a module) shouldn't an output pin not be allowed to be referenced with a sink expression?
